src/components/pagination/index.tsx

```diff
@@ -0,0 +1,270 @@
+			{start} – {end}
+			{showTotalItems ? ` of ${pagination.totalItems}` : null}
+		</div>
+	)
+}
+
+const Nav = ({ type = 'compact' }: NavProps) => {
+	const pagination = usePagination()
+
+	const totalPages = pagination.totalPages
+	const currentPage = pagination.page
+
+	const rawitems = useMemo(
+		() => Array.from({ length: totalPages }, (_, i) => i + 1),
+		[totalPages]
+	)
+
+	const items: (number | 'ellipsis')[] = ((t: NavProps['type']) => {
+		switch (t) {
+			case 'numbered':
+				return rawitems
+			case 'truncated':
+				return generateTruncatedList(rawitems, currentPage)
+			case 'compact':
+			default:
+				return []
+		}
+	})(type)
+
+	return (
+		<nav className={s.nav} aria-label="Pagination navigation">
+			<ButtonArrow type={type} direction="prev" />
+			{type !== 'compact' && (
+				<ul className={s['page-list']}>
+					{items.map((e, i) => {
+						if (e === 'ellipsis') {
+							// This key will be different on every re-render which is OK
+							const key = `${e}-${i}`
+							return (
+								<li className={classNames(s['page-item'])} key={key}>
+									<div className={s.ellipsis}>...</div>
+								</li>
+							)
+						}
+
+						const isSelected = e === currentPage
+						return (
+							<li
+								className={classNames(s['page-item'], {
+									[s['is-selected']]: isSelected,
+								})}
+								// This key will be stable between re-renders
+								// which prevents react from removing focus from
+								// the previous `document.activeElement`
+								key={e}
+							>
+								<ButtonNumber page={e}>{e}</ButtonNumber>
+							</li>
+						)
+					})}
+				</ul>
+			)}
+			<ButtonArrow type={type} direction="next" />
+		</nav>
+	)
+}
+
+const ButtonArrow = ({ type, direction }: ButtonArrowProps) => {
+	const pagination = usePagination()
+	const handleClick = {
+		next: () => {
+			// clamp upper bound
+			const nextIdx = Math.min(pagination.page + 1, pagination.totalPages)
+			pagination.setPage(nextIdx)
+			pagination.onPageChange(nextIdx)
+		},
+		prev: () => {
+			// clamp lower bound (0)
+			const prevIdx = Math.max(pagination.page - 1, 0)
+			pagination.setPage(prevIdx)
+			pagination.onPageChange(prevIdx)
+		},
+	}[direction]
+
+	const icon = {
+		next: <IconChevronRight16 />,
+		prev: <IconChevronLeft16 />,
+	}[direction]
+
+	const ariaLabel = {
+		next: 'Next page',
+		prev: 'Previous page',
+	}[direction]
+
+	const label = {
+		next: 'Next',
+		prev: 'Previous',
+	}[direction]
+
+	const isDisabled = {
+		next: pagination.page === pagination.totalPages,
+		prev: pagination.page === 1,
+	}[direction]
+
+	return (
+		<button
+			className={classNames(
+				s.arrow,
+				s.control,
+				{
+					next: s.next,
+					prev: s.prev,
+				}[direction]
+			)}
+			aria-label={ariaLabel}
+			onClick={handleClick}
+			disabled={isDisabled}
+		>
+			{icon}
+			{type === 'compact' && (
+				<span className={classNames(s.label)} aria-hidden="true">
+					{label}
+				</span>
+			)}
+		</button>
+	)
+}
```

Comment:
🚧 A few things for `ButtonArrow`:

- For the voice control experience to work, the Accessible Name of the rendered `<button>` has to match the visible label of it. In this case, where we're setting the Accessible Name `aria-label`, that means `aria-label` needs to be the same as the visible label (`label`).
- Rather than several maps with `prev/next` properties, we can use just one.
- The `{ next, prev }` map that sets a `className` isn't needed and can be simplified with `s[direction]`.
- `aria-hidden` should be set on the `span`. It's redundant because `aria-label` is being set on the element that contains it.

```tsx
const ButtonArrow = ({ type, direction }: ButtonArrowProps) => {
	const pagination = usePagination()

	const { handleClick, icon, isDisabled, label } = {
		next: {
			handleClick: () => {
				// clamp upper bound
				const nextIdx = Math.min(pagination.page + 1, pagination.totalPages)
				pagination.setPage(nextIdx)
				pagination.onPageChange(nextIdx)
			},
			icon: <IconChevronRight16 />,
			isDisabled: pagination.page === pagination.totalPages,
			label: 'Next',
		},
		prev: {
			handleClick: () => {
				// clamp lower bound (0)
				const prevIdx = Math.max(pagination.page - 1, 0)
				pagination.setPage(prevIdx)
				pagination.onPageChange(prevIdx)
			},
			icon: <IconChevronLeft16 />,
			isDisabled: pagination.page === 1,
			label: 'Previous',
		},
	}[direction]

	return (
		<button
			aria-label={label}
			className={classNames(s.arrow, s.control, s[direction])}
			disabled={isDisabled}
			onClick={handleClick}
		>
			{icon}
			{type === 'compact' && (
				<span className={classNames(s.label)}>{label}</span>
			)}
		</button>
	)
}
```

Reply:
Handled in https://github.com/hashicorp/dev-portal/pull/1508/commits/3af6b68814714e2837b4a3caedfebb88b3ce7f03
- re `s[direction]` — I opted for a `className` field that maps to `s.prev | s.next` . I didn't want to _mask_ a css-module class, behind a dynamic key (`direction`) in TSX, despite them _implicitly_ matching. I opted to be explicit here.

Reply:
Removed redundant `aria-hidden` https://github.com/hashicorp/dev-portal/pull/1508/commits/15630c2c8eff24e4d4267ad7c5ce35d2fef47de1